src/Cli/dotnet/Installer/Windows/MsiPackageCache.cs

```diff
@@ -0,0 +1,240 @@
+            {
+                throw new FileNotFoundException($"Manifest file not found: {manifestPath}");
+            }
+
+            Elevate();
+
+            if (IsElevated)
+            {
+                CreateRootDirectory();
+
+                string packageDirectory = GetPackageDirectory(packageId, packageVersion);
+
+                // Delete the directory and create a new one that's secure. If the files were properly
+                // cached, the client won't request this action.
+                if (Directory.Exists(packageDirectory))
+                {
+                    Directory.Delete(packageDirectory, recursive: true);
+                }
+
+                CreateSecureDirectory(packageDirectory);
+
+                // We cannot assume that the MSI adjacent to the manifest is the one to cache. We'll trust
+                // the manifest to provide the MSI filename.
+                MsiManifest msiManifest = JsonConvert.DeserializeObject<MsiManifest>(File.ReadAllText(manifestPath));
+                // Only use the filename+extension of the payload property in case the manifest has been altered.
+                string msiPath = Path.Combine(Path.GetDirectoryName(manifestPath), Path.GetFileName(msiManifest.Payload));
+
+                string cachedMsiPath = Path.Combine(packageDirectory, Path.GetFileName(msiPath));
+                string cachedManifestPath = Path.Combine(packageDirectory, Path.GetFileName(manifestPath));
+
+                MoveFile(manifestPath, cachedManifestPath);
+                MoveFile(msiPath, cachedMsiPath);
+            }
+            else if (IsClient)
+            {
+                Dispatcher.SendCacheRequest(InstallRequestType.CachePayload, manifestPath, packageId, packageVersion);
+            }
+        }
+
+        /// <summary>
+        /// Gets the full path of the cache directory for the specified package ID and version.
+        /// </summary>
+        /// <param name="packageId">The ID of the MSI workload pack package.</param>
+        /// <param name="packageVersion">The version of the MSI workload pack package.</param>
+        /// <returns>The directory where the MSI package will be cached.</returns>
+        public string GetPackageDirectory(string packageId, string packageVersion)
+        {
+            return Path.Combine(PackageCacheRoot, packageId, packageVersion);
+        }
+
+        /// <summary>
+        /// Moves a file from one location to another if the destination file does not already exist.
+        /// </summary>
+        /// <param name="sourceFile">The source file to move.</param>
+        /// <param name="destinationFile">The destination where the source file will be moved.</param>
+        protected void MoveFile(string sourceFile, string destinationFile)
+        {
+            if (!File.Exists(destinationFile))
+            {
+                FileAccessRetrier.RetryOnMoveAccessFailure(() => File.Move(sourceFile, destinationFile));
```

Comment:
I believe File.Move will error when moving across mounts, is there a situation where that will be a problem here?